rust/src/ssh/ssh.rs

```diff
@@ -0,0 +1,629 @@
+/* Copyright (C) 2018 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::parser;
+use crate::applayer::{LoggerFlags, TxDetectFlags};
+use crate::core::STREAM_TOSERVER;
+use crate::core::{self, AppProto, Flow, ALPROTO_UNKNOWN, IPPROTO_TCP};
+use crate::log::*;
+use crate::parser::*;
+use std;
+use std::ffi::{CStr, CString};
+use std::mem::transmute;
+
+static mut ALPROTO_SSH: AppProto = ALPROTO_UNKNOWN;
+
+#[repr(u32)]
+pub enum SSHEvent {
+    InvalidBanner = 0,
+    LongBanner,
+    InvalidRecord,
+}
+
+impl SSHEvent {
+    fn from_i32(value: i32) -> Option<SSHEvent> {
+        match value {
+            0 => Some(SSHEvent::InvalidBanner),
+            1 => Some(SSHEvent::LongBanner),
+            2 => Some(SSHEvent::InvalidRecord),
+            _ => None,
+        }
+    }
+}
+
+pub enum SSHConnectionState {
+    SshStateInProgress,
+    SshStateBannerDone,
+    SshStateFinished,
+}
+
+bitflags! {
+    pub struct SSHTxFlag: u8 {
+        const SSH_FLAG_NONE = 0;
+        const SSH_FLAG_VERSION_PARSED=1;
+        const SSH_FLAG_PARSER_DONE=2;
+    }
+}
+
+const SSH_MAX_BANNER_LEN: usize = 256;
+const SSH_RECORD_HEADER_LEN: usize = 6;
+//TODO complete enum and parse messages contents
+const SSH_MSG_NEWKEYS: u8 = 21;
+
+pub struct SshHeader {
+    record_left: u32,
+    record_buf: Vec<u8>,
+    flags: SSHTxFlag,
+    banner: Vec<u8>,
+    //can we have these be references to banner ?
```

Comment:
As the comment states : can we have these be slices referencing the banner ? (or is it too dangerous ?)

Reply:
It can be done. It will be tricky due to Rusts policing of holding multiple references

Reply:
I tried and it seemed tricky indeed.
So, should it be done ? Is it worth the performance cost ?

Reply:
Perf impact is probably low, although in general we don't know for sure until we test/measure. But I'm fine with leaving this as is for now.